lib/client-retry.js

```diff
@@ -0,0 +1,170 @@
+const TOO_MANY_REQUEST_ERROR = 'TooManyRequestsError';
+const TOO_MANY_REQUEST_ERROR_CODE = 429;
+/**
+ * Small wrapper around AssetComputeClient adding retry logic on 429s based on `retry-after` header
+ */
+class AssetComputeClientWithRetry extends AssetComputeClient {
+    constructor(integration, options) {
+        super(integration, options);
+        this.max429RetryCount = (options && options.max429RetryCount) || MAX_RETRY_ON_HTTP_429;
+    }
+
+    /**
+     * Determine time in milliseconds to wait until retrying
+     * @param {number} retryAfter optional amount in seconds to wait for until retrying
+     * @returns waitTime in milliseconds to wait before retrying
+     */
+    retryWaitTime(retryAfter) {
+        // randomly choose between 30-60 seconds waittime
+        let waitTime = Math.floor(Math.random() * (MAX_RETRY_WAIT_TIME_MS - MIN_RETRY_WAIT_TIME_MS + 1)) + MIN_RETRY_WAIT_TIME_MS;
+
+        // API gateway `retry-after` header is the amount of time in seconds to wait before retrying
+        if (retryAfter && typeof(retryAfter) === 'number' && !isNaN(retryAfter)) {
+            waitTime = retryAfter * 1000 + (Math.floor(Math.random() * 1000));
+        }
+        return waitTime;
+    }
+
+    /**
+     * Determine if should retry on 429
+     * 
+     * It will continue to retry if it has not received anything other than a 429 response 
+     * or retry count is up
+     * @param {boolean} non429Response false if received a 429 response
+     * @param {number} retryCount 
+     * @returns 
+     */
+    shouldRetry(non429Response, retryCount) {
+        return !non429Response && retryCount <= this.max429RetryCount;
+    }
+
+    async register() {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (this.shouldRetry(non429Response, retryCount)) {
+            retryCount++;
+            try {
+                response = await super.register();
+                non429Response = true;
+            } catch (e) {
+                // string comparision instead of `instanceof` since `instanceof` causes issues
+                if (e.code === TOO_MANY_REQUEST_ERROR_CODE && e.name === TOO_MANY_REQUEST_ERROR) {
+                    non429Response = false;
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e.retryAfter);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+        if (!non429Response) {
+            throw new Error(`Running into 429s after ${retryCount} attempts. Will not continue retrying. Try again in a few minutes.`);
+
+        }
+        // pass through error
+        if (error) {
+            throw error;
+        }
+        return response;
```

Comment:
I think this may be easier to understand if you track `lastError`, set when caught and create a function that could tell you if an error is a `Too Many Request` error. Then replace the end with something like:

```
if (!lastError) {
return response;
}else if (tooManyRequestError(lastError)) {
throw new Error(`Running in to 429s...`);
} else {
throw lastError;
}
```